### PR TITLE
Query: Allow using STET overload for Set in query filter/defining query

### LINKED ISSUE #24601: Defining query/query filter throws when using shared type entity type Set overload
```
modelBuilder.Entity<AnimalQuery>().ToInMemoryQuery(() => context.Set<Dictionary<string, object>>("BirdSTET").Select(b => MaterializeView(b)));
```
```
  Message: 
    System.ArgumentNullException : Value cannot be null. (Parameter 'entityType')
  Stack Trace: 
    Check.NotNull[T](T value, String parameterName) line 27
    QueryRootExpression.ctor(IEntityType entityType) line 45
    DbSetAccessRewritingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression) line 114
    MethodCallExpression.Accept(ExpressionVisitor visitor)
    ExpressionVisitor.Visit(Expression node)
    ExpressionVisitor.VisitLambda[T](Expression`1 node)
    Expression`1.Accept(ExpressionVisitor visitor)
    ExpressionVisitor.Visit(Expression node)
    DbSetAccessRewritingExpressionVisitor.Rewrite(IReadOnlyModel model, Expression expression) line 81
    DefiningQueryRewritingConvention.ProcessModelFinalizing(IConventionModelBuilder modelBuilder, IConventionContext`1 context) line 38
    ImmediateConventionScope.OnModelFinalizing(IConventionModelBuilder modelBuilder) line 74
    ConventionDispatcher.OnModelFinalizing(IConventionModelBuilder modelBuilder) line 64
    Model.FinalizeModel() line 858
    ModelBuilder.FinalizeModel() line 547
    ModelSource.CreateModel(DbContext context, IConventionSetBuilder conventionSetBuilder, ModelDependencies modelDependencies) line 181
    ModelSource.GetModel(DbContext context, ModelCreationDependencies modelCreationDependencies, Boolean designTime) line 130
    DbContextServices.CreateModel(Boolean designTime) line 87
    DbContextServices.get_Model() line 114
    <>c.<TryAddCoreServices>b__7_3(IServiceProvider p) line 256
    CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.ResolveService(ServiceCallSite callSite, RuntimeResolverContext context, RuntimeResolverLock lockType, ServiceProviderEngineScope serviceProviderEngine)
    CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
    CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite callSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.ResolveService(ServiceCallSite callSite, RuntimeResolverContext context, RuntimeResolverLock lockType, ServiceProviderEngineScope serviceProviderEngine)
    CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
    CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite callSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.ResolveService(ServiceCallSite callSite, RuntimeResolverContext context, RuntimeResolverLock lockType, ServiceProviderEngineScope serviceProviderEngine)
    CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
    CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite callSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.ResolveService(ServiceCallSite callSite, RuntimeResolverContext context, RuntimeResolverLock lockType, ServiceProviderEngineScope serviceProviderEngine)
    CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
    CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite callSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.ResolveService(ServiceCallSite callSite, RuntimeResolverContext context, RuntimeResolverLock lockType, ServiceProviderEngineScope serviceProviderEngine)
    CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
    CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite callSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.ResolveService(ServiceCallSite callSite, RuntimeResolverContext context, RuntimeResolverLock lockType, ServiceProviderEngineScope serviceProviderEngine)
    CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
    CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite callSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
    <>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
    ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
    ServiceProviderEngineScope.GetService(Type serviceType)
    ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
    ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
    DbContext.get_DbContextDependencies() line 399
    DbContext.get_ContextServices() line 381
    DbContext.get_InternalServiceProvider() line 335
    IServiceProvider>.get_Instance() line 1801
```

### PR DESCRIPTION
**Pull request by smitpatel**
Resolves #24601

### CODE REVIEW

File: test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs [hunk 1]
```diff
@@ -10229,6 +10229,55 @@ public Test24569(int value)
 
         #endregion
 
+        #region Issue24601
+
+        [ConditionalFact]
+        public virtual async Task Can_use_shared_type_entity_type_in_query_filter()
```
Comment (AndriySvyryd) [inline test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs:10235]: Don't add tests to QueryBugsTest

Comment (smitpatel) [reply test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs:10235]: Which testbase contains tests for a custom model?

Comment (AndriySvyryd) [reply test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs:10235]: NonSharedModelTestBase

Comment (AndriySvyryd) [reply test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs:10235]: You can create a new one for this test - `SharedTypeQueryTestBase`

**Comment by smitpatel** [comment]
@ajcvickers @AndriySvyryd - Do you want this to go to rc1?

**Comment by ajcvickers** [comment]
@smitpatel Yep.

File: src/EFCore/Metadata/Conventions/QueryFilterRewritingConvention.cs [hunk 2]
```diff
@@ -111,7 +114,55 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
...
+
+                        throw new InvalidOperationException(message);
+                    }
+
+                    if (entityType.ClrType != returnType)
+                    {
+                        var message = CoreStrings.DbSetIncorrectGenericType(
+                            entityType.ShortName(), entityType.ClrType.ShortDisplayName(), returnType.ShortDisplayName());
+
+                        throw new InvalidOperationException(message);
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Conventions/QueryFilterRewritingConvention.cs:162]: Add tests for the new exceptions

**Review by AndriySvyryd — APPROVED** [review]

**Pull request merged**